tsdate/core.py

```diff
@@ -858,6 +931,129 @@ def outside_maximization(self, *, eps, progress=None):
+        assert self.priors.timepoints.size == 2
+
+        # Messages passed from factors in the direction of roots
+        self.parent_message = np.tile(
+            self.lik.identity_constant,
+            (self.ts.num_edges, 1),
+        )
+
+        # Messages passed from factors in the direction of leaves
+        self.child_message = np.tile(
+            self.lik.identity_constant,
+            (self.ts.num_edges, 1),
+        )
+
+        # Normalizing constants from each factor
+        self.factor_norm = np.full(self.ts.num_edges, 0.0)
+
+        # the approximate posterior marginals
+        self.posterior = self.priors.clone_with_new_data(
+            grid_data=self.priors.grid_data.copy(),
+            fixed_data=np.nan,
+        )
+
+        # factors for edges leading from fixed nodes are invariant
+        # and can be incorporated into the posterior beforehand
+        for edge in self.ts.edges():
+            if edge.child in self.fixednodes:
+                self.parent_message[edge.id] = self.lik.to_gamma(edge)
+                self.posterior[edge.parent] = self.lik.combine(
+                    self.posterior[edge.parent], self.parent_message[edge.id]
+                )
+                # self.factor_norm[edge.id] += ... # TODO
+
+    def edges_by_parent_asc(self):
+        """
+        Edges in order of increasing age of parent
+        """
+        return self.ts.edges()
+
+    def edges_by_child_desc(self):
+        """
+        Edges in order of decreasing age of child
+        """
+        wtype = np.dtype(
+            [
+                ("child_age", self.ts.tables.nodes.time.dtype),
+                ("child_node", self.ts.tables.edges.child.dtype),
+            ]
+        )
+        w = np.empty(self.ts.num_edges, dtype=wtype)
+        w["child_age"] = self.ts.tables.nodes.time[self.ts.tables.edges.child]
+        w["child_node"] = self.ts.tables.edges.child
+        sorted_child_parent = (
+            self.ts.edge(i)
+            for i in reversed(np.argsort(w, order=("child_age", "child_node")))
+        )
+        return sorted_child_parent
+
+    def propagate(self, *, edges, progress=None):
+        """
+        Update approximating factor for each edge
+        """
+        if progress is None:
+            progress = self.progress
+        # TODO: this will still converge if parallelized (potentially slower)
+        for edge in tqdm(
+            edges,
+            desc="Expectation Propagation",
+            total=self.ts.num_edges,
+            disable=not progress,
+        ):
+            if edge.child in self.fixednodes:
+                continue
+            if edge.parent in self.fixednodes:
+                raise ValueError("Internal nodes can not be fixed in EP algorithm")
+            edge_lik = self.lik.to_gamma(edge)
+            edge_lik += np.array([-1.0, 0.0])  # to Poisson, TODO cleanup
+            # cavity posteriors without approximate edge factor
```

Comment:
What is the cavity posterior. Is there an explanation somewhere?

Reply:
The idea is that you can remove the contribution of an edge to the global approximation by subtracting natural parameters. The resulting reduced approximation is called the "cavity" (for a particular edge) in the BP/EP literature. Then, moment matching is performed against the cavity + the exact edge likelihood, to get a new "approximate edge factor".

I can add some description to the docstrings.